### PR TITLE
fix(N-02): use custom errors

### PR DESCRIPTION
**Pull request by DinshawKothari**
Ticket: USDS-128

### CODE REVIEW

File: contracts/USDS.sol [hunk 1]
```diff
@@ -31,6 +31,21 @@ contract USDS is
     bytes32 public constant UPGRADER_ROLE = keccak256("UPGRADER_ROLE");
     bytes32 public constant RESCUER_ROLE = keccak256("RESCUER_ROLE");
 
+    // --- Custom Errors ---
+    error InvalidAddress();
```
Comment (sachushaji) [inline contracts/USDS.sol:45]: Can you reorder this according to the style guide https://docs.soliditylang.org/en/v0.8.28/style-guide.html

Errors comes after events

File: contracts/USDS.sol [hunk 2]
```diff
@@ -193,9 +208,9 @@ contract USDS is
         uint256 value
     ) public virtual override returns (bool) {
         address spender = _msgSender();
-        require(!isBlacklisted(spender), "Spender is blacklisted");
-        require(!isBlacklisted(from), "Sender is blacklisted");
-        require(value != 0, "Amount must be > 0");
+        if (isBlacklisted(_msgSender())) revert SpenderBlacklisted();
+        if (isBlacklisted(from)) revert SenderBlacklisted();
+        if (value <= 0) revert InvalidAmount();
         _spendAllowance(from, spender, value);
```
Comment (sachushaji) [inline contracts/USDS.sol:214]:
```suggestion
        _spendAllowance(from, _msgSender(), value);
```

File: contracts/USDS.sol [hunk 3]
```diff
@@ -214,7 +229,7 @@ contract USDS is
         uint256 value
     ) public virtual override returns (bool) {
         address owner = _msgSender();
-        require(!isBlacklisted(owner), "Sender is blacklisted");
+        if (isBlacklisted(_msgSender())) revert SenderBlacklisted();
         _transfer(owner, to, value);
```
Comment (sachushaji) [inline contracts/USDS.sol:233]:
```suggestion
        _transfer(_msgSender(), to, value);
```

File: contracts/USDS.sol [hunk 4]
```diff
@@ -214,7 +229,7 @@ contract USDS is
         uint256 value
     ) public virtual override returns (bool) {
         address owner = _msgSender();
```
Comment (sachushaji) [inline contracts/USDS.sol:231]:
```suggestion
```

File: contracts/USDS.sol [hunk 5]
```diff
@@ -376,12 +377,9 @@ contract USDS is
         // does not exceed the available `reserves`.
         // In non-batch mode, the `mintAmount` is not yet included in `totalSupply`,
         // so we need to ensure that `totalSupply + mintAmount` stays within `reserves`.
-        require(
-            isBatch
-                ? currentSupply <= reserves
-                : currentSupply + mintAmount <= reserves,
-            "Supply exceeds reserves"
-        );
+        if (isBatch ? currentSupply > reserves : currentSupply + mintAmount > reserves) {
```
Comment (sachushaji) [inline contracts/USDS.sol]:
```suggestion
        if (isBatch ? currentSupply > reserves : (currentSupply + mintAmount) > reserves) {
```

**Review by sachushaji — COMMENTED** [review]
Small changes, otherwise LGTM, one of the comments was to remove these temporary variables, I commented on some but if you are going to resolve that in a dedicated PR you can ignore those comments

File: contracts/USDS.sol (hunk 2 above)
Comment (DinshawKothari) [reply contracts/USDS.sol:214]: I have this addressed in separate PR => N-08 Reimplemented Transfer Logic

File: contracts/USDS.sol (hunk 3 above)
Comment (DinshawKothari) [reply contracts/USDS.sol:233]: I have this addressed in separate PR => N-08 Reimplemented Transfer Logic

File: contracts/USDS.sol (hunk 4 above)
Comment (DinshawKothari) [reply contracts/USDS.sol:231]: I have this addressed in separate PR => N-08 Reimplemented Transfer Logic

File: contracts/USDS.sol (hunk 1 above)
Comment (DinshawKothari) [reply contracts/USDS.sol:45]: done

File: contracts/USDS.sol (hunk 5 above)
Comment (DinshawKothari) [reply contracts/USDS.sol]: done

**Review by sachushaji — APPROVED** [review]
LGTM!!, Keep it coming!!

**Pull request merged**